File: C#0.7/Nabla.cs
using System;

namespace MyMath.Vectors.VectorCalculus
{
    public class Nabla
    {
        //No way to init as object
        private Nabla() { }

        public static Func<Vector, double> Divergence(Func<Vector, Vector> v, double dx)
        {
            double res(Vector a)
            {
                double x = (v(Vector.Add(a, new Vector(dx, 0, 0))).x - v(Vector.Add(a, new Vector(-dx, 0, 0))).x) / (2 * dx);
                double y = (v(Vector.Add(a, new Vector(0, dx, 0))).y - v(Vector.Add(a, new Vector(0, -dx, 0))).y) / (2 * dx);
                double z = (v(Vector.Add(a, new Vector(0,  0, dx))).z - v(Vector.Add(a, new Vector(0, 0, -dx))).z) / (2 * dx);
                return x + y + z;

            }
            return res;

        }
        public static Func<Vector, Vector> Curl(Func<Vector, Vector> v, double dx)
        {
            Vector res(Vector a)
            {
                Vector x(Vector c)
                {
                    return Vector.VectorProd(v(c), Vector.I);

                }
                Vector y(Vector c)
                {
                    return Vector.VectorProd(v(c), Vector.J);

                }
                Vector z(Vector c)
                {
                    return Vector.VectorProd(v(c), Vector.K);

                }
                return Vector.Add(Vector.Add(VectorDer.derivX(x, a, dx), VectorDer.derivY(y, a, dx)), VectorDer.derivZ(z, a, dx));

            }
            return res;

        }
        public static Func<Vector, Vector> Gradient(Func<Vector, double> f, double dx)
        {
            Vector res(Vector a)
            {
                double x()
                {
                    double f1 = f(Vector.Add(a, new Vector(dx, 0, 0)));
                    double f2 = f(Vector.Add(a, new Vector(-dx, 0, 0)));
                    return (f1 - f2) / (2 * dx);

                }
                double y()
                {
                    double f1 = f(Vector.Add(a, new Vector(0, dx, 0)));
                    double f2 = f(Vector.Add(a, new Vector(0, -dx, 0)));
                    return (f1 - f2) / (2 * dx);

                }
                double z()
                {
                    double f1 = f(Vector.Add(a, new Vector(0, 0, dx)));
                    double f2 = f(Vector.Add(a, new Vector(0, 0, -dx)));
                    return (f1 - f2) / (2 * dx);

                }
                return new Vector(x(), y(), z());

            }
            return res;

        }

    }

}

File: 2.1.0/JGJAR/libraries/MyMath/vectors/vectorCalculus/Nabla.java
package JGJAR.libraries.MyMath.vectors.vectorCalculus;

import JGJAR.libraries.MyMath.vectors.Vector;
import JGJAR.libraries.MyMath.vectors.VectorFunction;

/**<h1>Nabla class</h1>
 * 
 * @author JavaGamesJAR
 * @version 1.0
 * @since 2.0.0
 *
 */
public class Nabla 
{
	/**Returns curl (&nabla;&Cross;v) of a function
	 * 
	 * @param f function
	 * @return curl f
	 * @since 1.0 (MyMath 2.0.0)
	 */
	public static VectorFunction curl(VectorFunction f)
	{
		VectorFunction x = new VectorFunction() {
			
			@Override
			public Vector apply(Vector a) {
				return Vector.VectorProduct(f.apply(a), Vector.I);
			}
		};
		VectorFunction y = new VectorFunction() {
			
			@Override
			public Vector apply(Vector a) {
				return Vector.VectorProduct(f.apply(a), Vector.J);
			}
		};
		VectorFunction z = new VectorFunction() {
			
			@Override
			public Vector apply(Vector a) {
				return Vector.VectorProduct(f.apply(a), Vector.K);
			}
		};
		VectorFunction dx = VectorDerivative.easyDerivativeX(x);
		VectorFunction dy = VectorDerivative.easyDerivativeY(y);
		VectorFunction dz = VectorDerivative.easyDerivativeZ(z);
		
		return new VectorFunction() {
			
			@Override
			public Vector apply(Vector a) {
				return Vector.add(dx.apply(a), dy.apply(a), dz.apply(a));
			}
		};
		
	}
	/**Returns divergence (&nabla;&sdot;v) of a function
	 * 
	 * @param f function
	 * @return div f
	 * @since 1.0 (MyMath 2.0.0)
	 */
	public static ScalarAtPointFunction div(VectorFunction f)
	{
		final ScalarAtPointFunction x = new ScalarAtPointFunction() {
			
			@Override
			public double apply(Vector a) {
				return f.apply(a).x;
			}
		};
		final ScalarAtPointFunction y = new ScalarAtPointFunction() {
			
			@Override
			public double apply(Vector a) {
				return f.apply(a).y;
			}
		};
		final ScalarAtPointFunction z = new ScalarAtPointFunction() {
			
			@Override
			public double apply(Vector a) {
				return f.apply(a).z;
			}
		};
		return new ScalarAtPointFunction() {
			
			@Override
			public double apply(Vector a) {
				return ScalarAtPointFunction.derivX(x).apply(a)+
						ScalarAtPointFunction.derivY(y).apply(a)+ScalarAtPointFunction.derivZ(z).apply(a);
			}
		};
		
	}
	/**Returns gradient (&nabla;&phi;) of a function
	 * 
	 * @param f function
	 * @return grad f
	 * @since 1.0 (MyMath 2.0.0)
	 */
	public static VectorFunction grad(ScalarAtPointFunction f)
	{
		return new VectorFunction() {
			
			@Override
			public Vector apply(Vector a) {
				return new Vector(ScalarAtPointFunction.derivX(f).apply(a), 
						ScalarAtPointFunction.derivY(f).apply(a), ScalarAtPointFunction.derivZ(f).apply(a));
			}
		};
		
	}

}
